pkg/registrars/registration/ampcache-registrar.go

```diff
@@ -0,0 +1,294 @@
+	req, err := r.executeRequestCommon(ctx, payload, logger)
+	if err != nil {
+		return err
+	}
+	resp, err := r.client.Do(req)
+	if err != nil {
+		logger.Warnf("%v failed to do HTTP request to registration endpoint %s: %v", r.endpoint, err)
+		return err
+	}
+	logger.Printf("Response: %v", resp)
+	defer resp.Body.Close()
+
+	logger.Printf("AMP cache rendezvous response: %s", resp.Status)
+
+	// Check that the HTTP request returned a success code
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		// logger.Warnf("got non-success response code %d from registration endpoint %v", resp.StatusCode, r.endpoint)
+		return fmt.Errorf("non-success response code %d on %s", resp.StatusCode, r.endpoint)
+	}
+
+	if _, err := resp.Location(); err == nil {
+		// The Google AMP Cache may return a "silent redirect" with
+		// status 200, a Location header set, and a JavaScript redirect
+		// in the body. The redirect points directly at the origin
+		// server for the request (bypassing the AMP cache). We do not
+		// follow redirects nor execute JavaScript, but in any case we
+		// cannot extract information from this response and can only
+		// treat it as an error.
+		return fmt.Errorf("non-success silent redirect %d on %s", resp.StatusCode, r.endpoint)
+	}
+	return nil
+}
+
+func (r AMPCacheRegistrar) executeAMPCacheRequestBidirectional(ctx context.Context, payload []byte, logger logrus.FieldLogger) (*pb.RegistrationResponse, error) {
+	// Create an instance of the ConjureReg struct to return; this will hold the updated phantom4 and phantom6 addresses received from registrar response
+	regResp := &pb.RegistrationResponse{}
+	req, err := r.executeRequestCommon(ctx, payload, logger)
+	if err != nil {
+		return regResp, err
+	}
+	resp, err := r.client.Do(req)
+	if err != nil {
+		logger.Warnf("%v failed to do HTTP request to registration endpoint %s: %v", r.endpoint, err)
+		return nil, err
+	}
+	logger.Printf("Response: %v", resp)
+	defer resp.Body.Close()
+
+	logger.Printf("AMP cache rendezvous response: %s", resp.Status)
+
+	// Check that the HTTP request returned a success code
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		// logger.Warnf("got non-success response code %d from registration endpoint %v", resp.StatusCode, r.endpoint)
+		return nil, fmt.Errorf("non-success response code %d on %s", resp.StatusCode, r.endpoint)
+	}
+
+	if _, err := resp.Location(); err == nil {
+		// The Google AMP Cache may return a "silent redirect" with
+		// status 200, a Location header set, and a JavaScript redirect
+		// in the body. The redirect points directly at the origin
+		// server for the request (bypassing the AMP cache). We do not
+		// follow redirects nor execute JavaScript, but in any case we
+		// cannot extract information from this response and can only
+		// treat it as an error.
+		return nil, fmt.Errorf("non-success silent redirect %d on %s", resp.StatusCode, r.endpoint)
+	}
+	lr := io.LimitReader(resp.Body, readLimit+1)
```

Comment:
There is a feature when the client has a older `ClientConf` version, the server sends the updated one via the protobuf message. This currently causes client-side registration to fail with 
```
error sending request to Bidirectional AMPCache registrar: ReadAll failed with error: &{%!d(string=missing </pre> tag)}
```

The ClientConf size is ~108KB, and it probably won't be increased significantly. Increasing `readLimit` to 200KB fixed it in my test. 

Should we go ahead with this limit, or are there other factors we should consider? Also what's purpose of the limit?

Reply:
Yeah this change should be totally fine to do. That existing 100KB limit was inherited from a safeguard to prevent large reads in Snowflake because our rendezvous messages are very small (see https://gitlab.torproject.org/tpo/anti-censorship/pluggable-transports/snowflake/-/issues/26348). We could also remove it or increase it even more, if you think it's likely that the ClientConf will grow significantly. Since it's a client-side value it will be more complicated to update so we want to make sure that we give plenty of room.

Reply:
Gotcha, thanks for the context. I increased it to 1MB in case of upgrades, should be plenty of headroom